lindi/LindiH5pyFile/LindiH5pyFile.py

```diff
@@ -0,0 +1,218 @@
+            else:
+                with open(rfs, "r") as f:
+                    data = json.load(f)
+                assert isinstance(data, dict)  # prevent infinite recursion
+                return LindiH5pyFile.from_reference_file_system(data)
+        elif isinstance(rfs, dict):
+            fs = ReferenceFileSystem(rfs).get_mapper(root="")
+            return LindiH5pyFile.from_zarr_store(fs)
+        else:
+            raise Exception(f"Unhandled type for rfs: {type(rfs)}")
+
+    @staticmethod
+    def from_zarr_store(zarr_store: Union[ZarrStore, FSMap]):
+        """
+        Create a LindiH5pyFile from a zarr store.
+        """
+        zarr_group = zarr.open(store=zarr_store, mode="r")
+        assert isinstance(zarr_group, zarr.Group)
+        return LindiH5pyFile.from_zarr_group(zarr_group)
+
+    @staticmethod
+    def from_zarr_group(zarr_group: zarr.Group):
+        """
+        Create a LindiH5pyFile from a zarr group.
+        """
+        return LindiH5pyFile(zarr_group)
+
+    @staticmethod
+    def from_h5py_file(h5py_file: h5py.File):
+        """
+        Create a LindiH5pyFile from an h5py file.
+        """
+        return LindiH5pyFile(h5py_file)
+
+    @property
+    def attrs(self):  # type: ignore
+        if isinstance(self._file_object, h5py.File):
+            attrs_type = 'h5py'
+        elif isinstance(self._file_object, zarr.Group):
+            attrs_type = 'zarr'
+        else:
+            raise Exception(f'Unexpected file object type: {type(self._file_object)}')
+        return LindiH5pyAttributes(self._file_object.attrs, attrs_type=attrs_type)
+
+    @property
+    def filename(self):
+        # This is not a string, but this is what h5py seems to do
+        if isinstance(self._file_object, h5py.File):
+            return self._file_object.filename
+        elif isinstance(self._file_object, zarr.Group):
+            return ''
+        else:
+            raise Exception(f"Unhandled type for file object: {type(self._file_object)}")
+
+    @property
+    def driver(self):
+        raise Exception("Getting driver is not allowed")
+
+    # @property
+    # def mode(self):
+    #     if isinstance(self._file_object, h5py.File):
+    #         return self._file_object.mode
+    #     elif isinstance(self._file_object, zarr.Group):
+    #         # hard-coded to read-only
+    #         return "r"
+    #     else:
+    #         raise Exception(f"Unhandled type: {type(self._file_object)}")
+
+    @property
+    def libver(self):
+        raise Exception("Getting libver is not allowed")
+
+    @property
+    def userblock_size(self):
+        raise Exception("Getting userblock_size is not allowed")
+
+    @property
+    def meta_block_size(self):
+        raise Exception("Getting meta_block_size is not allowed")
+
+    def swmr_mode(self, value):  # type: ignore
+        raise Exception("Getting swmr_mode is not allowed")
+
+    def close(self):
+        # return self._file_object.close()
```

Comment:
Do we not want to close the h5py.File object?

Reply:
Well, it's not opened in the constructor, so I don't think it should be closed here.

Reply:
Good point. I missed that.

Reply:
`__init__` states that users should use the `from_*` factory methods instead. Who is responsible for closing objects that are being opened there? E.g., in `from_zarr_store` the factory methods opens the zarr store, but other than the returned `LindiH5pyFile` object, no-one else has access/owns the file.

https://github.com/NeurodataWithoutBorders/lindi/blob/fe51dbc3aafe7c2a35b392fc93059387e96319a6/lindi/LindiH5pyFile/LindiH5pyFile.py#L52-L59

 For factory methods that require an already opened file to be handed in, I agree, that LINIDI should not just close it. But for functions where LINDI functions open the file (and in particular when the opened file/store is not being exposed to the caller), I think it should be the role of LINDI to close.